### ISSUE TITLE
MkDocs and PyInstaller

### ISSUE THREAD

**Issue by SchulteMarkus** [mkdocs/mkdocs#2711]
I would love to compile MkDocs to binary using [PyInstaller](https://pyinstaller.readthedocs.io/en/stable/usage.html).

**tl;dr**: MkDocs compiles to an executable binary, but this does not find the (pre-)installed themes.

**Reproduce**

1. Installing MkDocs and PyInstaller using pip is straight forward

```bash
$ pip --version
pip 21.3.1 from /usr/local/lib/python3.8/dist-packages/pip (python 3.8)
$ mkdocs -V
mkdocs, version 1.2.3 from /home/ui/.local/lib/python3.8/site-packages/mkdocs (Python 3.8)
$ pyinstaller --version
4.7
```

2. I setup a minimized project as described on https://www.mkdocs.org/getting-started/

3. Building using MkDocs itself works fine
```bash
$ mkdocs build
INFO     -  Cleaning site directory
INFO     -  Building documentation to directory: my-project/site
INFO     -  Documentation built in 0.15 seconds
```

4. Compiling MkDocs to binary works fine
```bash
$ pyinstaller --onefile --name mkdocs mkdocs 
110 INFO: PyInstaller: 4.7
110 INFO: Python: 3.8.10
131 INFO: Platform: Linux-5.15.6-051506-generic-x86_64-with-glibc2.29
...
~/.local/bin$ dist/mkdocs -V
mkdocs, version 1.2.3 from /tmp/_MEIyugeg5/mkdocs (Python 3.8)
```

5. But using the MkDocs binary breaks:
```bash
$ ~/.local/bin/dist/mkdocs build
ERROR    -  Config value: 'theme'. Error: Unrecognised theme name: 'mkdocs'. The available installed themes are:
ERROR    -  Config value: 'plugins'. Error: The "search" plugin is not installed
Aborted with 2 Configuration Errors!
```

**Background**

I am not a Python dev, so I don't know if this is the right place to raise an issue or contact the PyInstaller guys. 
Anyhow, I appreciate any kind of help for getting MkDocs to binary!

**Comment by oprypin** [comment]
So the issue is all about setuptools entry_points.

https://github.com/mkdocs/mkdocs/blob/97eafb9094fdd56d19022ae8fda332e728df674b/setup.py#L75-L86

The same will also apply to markdown_extensions if you use any.

I am not familiar with pyinstaller (or entry_points, to be honest) but seems MkDocs absolutely requires support of this functionality of these things being "installed" or "registered" via setup.py. Then either pyinstaller can handle that or not (which it apparently cannot, or something went wrong, or maybe it's not even possible for it to handle that!)

**Comment by rokm** [comment]
You need to ensure that metadata is collected, that all indirect/lazy imported submodules are collected, and that non-python data is collected:
```
pyinstaller --clean --noconfirm venv/bin/mkdocs \
  --recursive-copy-metadata mkdocs \
  --collect-submodules mkdocs \
  --collect-data mkdocs \
  --onefile
```
or
```
pyinstaller --clean --noconfirm venv/bin/mkdocs \
  --recursive-copy-metadata mkdocs \
  --collect-submodules mkdocs \
  --collect-data mkdocs \
  --onedir --name mkdocs-bin
```
(for onedir, executable must not be named `mkdocs`, because it will clash with package's directory. That's not a problem in onefile build).

**Comment by SchulteMarkus** [comment]
> You need to ensure that metadata is collected, that all indirect/lazy imported submodules are collected, and that non-python data is collected:
> 
> ```
> pyinstaller --clean --noconfirm venv/bin/mkdocs \
>   --recursive-copy-metadata mkdocs \
>   --collect-submodules mkdocs \
>   --collect-data mkdocs \
>   --onefile
> ```
> 
> or
> 
> ```
> pyinstaller --clean --noconfirm venv/bin/mkdocs \
>   --recursive-copy-metadata mkdocs \
>   --collect-submodules mkdocs \
>   --collect-data mkdocs \
>   --onedir --name mkdocs-bin
> ```
> 
> (for onedir, executable must not be named `mkdocs`, because it will clash with package's directory. That's not a problem in onefile build).

Gsus, thanks for the fast response! Give it a try on Monday (CET).

**Comment by SchulteMarkus** [comment]
Works, awesome!

```bash
~/.local/bin$ which mkdocs
/home/ui/.local/bin/mkdocs
~/.local/bin$ pyinstaller --recursive-copy-metadata mkdocs --collect-submodules mkdocs --collect-data mkdocs --onefile mkdocs
...
10531 INFO: Building EXE from EXE-00.toc completed successfully.
~/tmp/test/my-project$ ~/.local/bin/dist/mkdocs build
INFO     -  Cleaning site directory
INFO     -  Building documentation to directory: /home/ui/tmp/test/my-project/site
INFO     -  Documentation built in 0.05 seconds
```

**Comment by Leroy-X** [comment]
So, where to download binary version? thanks
